### PR TITLE
feat: Add RedisTimeSeries plugin

### PR DESCRIPTION
**Pull request by gkorland**
### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

### CODE REVIEW

**Comment by gkorland** [comment]
@srebhan
Followup on  - https://github.com/influxdata/telegraf/pull/5275#pullrequestreview-673072568

-      The formatting of the sampleConfig and the corresponding update in the README.
Done.

-      The superfluous length check for metrics in Write().
Done

-      Check for address being actually set before constructing the client and report an error if it is not.
Done

-      A comment to my question regarding using a mock-package to test this plugin without requiring a real database instance.
I'm not familiar of RedisTimeSeries mock for Go.

-      Newly added comment regarding simplification in Write().
I'm not sure about the simplification but anyway it seems to me much less efficient to but copy the tags twice.

File: plugins/outputs/redistimeseries/redistimeseries.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+}
+func (r *RedisTimeSeries) Write(metrics []telegraf.Metric) error {
+	for _, m := range metrics {
+		now := m.Time().UnixNano() / 1000000 // in milliseconds
+		name := m.Name()
+
+		var tags []interface{}
+		for k, v := range m.Tags() {
+			tags = append(tags, k)
+			tags = append(tags, v)
```
Comment (srebhan) [inline plugins/outputs/redistimeseries/redistimeseries.go:70]:
```suggestion
			tags = append(tags, k, v)
```

File: plugins/outputs/redistimeseries/redistimeseries.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+		for fieldName, value := range m.Fields() {
+			key := name + "_" + fieldName
+
+			var addSlice []interface{}
+			addSlice = append(addSlice, "TS.ADD")
+			addSlice = append(addSlice, key)
+			addSlice = append(addSlice, now)
+			addSlice = append(addSlice, value)
+			addSlice = append(addSlice, tags...)
```
Comment (srebhan) [inline plugins/outputs/redistimeseries/redistimeseries.go:81]:
```suggestion
			addSlice := []interface{} {"TS.ADD", key, now, value}
			addSlice = append(addSlice, tags...)
```

File: plugins/outputs/redistimeseries/redistimeseries.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+			if err := r.client.Do(addSlice...).Err(); err != nil {
+				return fmt.Errorf("reattempting adding sample failed: %v", err)
```
Comment (srebhan) [inline plugins/outputs/redistimeseries/redistimeseries.go:84]:
```suggestion
				return fmt.Errorf("adding sample failed: %v", err)
```
You maybe also want to print the sample with a `r.Log.Debugf` to see what is going on in debugging.

**Review by srebhan — COMMENTED** [review]
Thanks for the nice update @gkorland! I do have only some style suggestions, but the PR looks good overall. There are two more things I want to ask you:
1. Please sync the README example with the `sampleConfig` to show all options to the user.
2. Can you please add a test for the serialization? You can _fake_ a client by introducing an interface with a `Do()` function and then compare the serialized format to the expected one. This would give us at least some testing in short-mode.

File: plugins/outputs/redistimeseries/README.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+```
+
+## Required parameters
+
+* `address`: The address of the RedisTimeSeries server
+
+## Optional parameters
+
+* `password`: The password to connect RedisTimeSeries server.
```
[the remark this answers is not in the document]
Comment (gkorland) [reply plugins/outputs/redistimeseries/README.md:21]: I think it's better to leave it there so users will have official comment about what is Required/Optional

[the remark this answers is not in the document]
Comment (srebhan) [reply plugins/outputs/redistimeseries/README.md:21]: Fine with me. The more the merrier. :-)

File: plugins/outputs/redistimeseries/redistimeseries.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+)
+
+const sampleConfig = `
+  ## The address of the RedisTimeSeries server.
+  address = "127.0.0.1:6379"
+
+  ## Redis ACL credentials
+  # username = ""
+  # password = ""
+  # database = 0
```
Comment (srebhan) [inline plugins/outputs/redistimeseries/redistimeseries.go:19]: I still don't see `username` and `database` reflected in the README file, `sampleConfig` and the README example should match as this is what the user sees when viewing your plugin in github. Am I missing something?

**Review by srebhan — COMMENTED** [review]
This only leaves us with synchronizing the `sampleConfig` to the README example...

File: plugins/outputs/redistimeseries/redistimeseries.go (hunk 5 above)
Comment (chayim) [reply plugins/outputs/redistimeseries/redistimeseries.go:19]: @srebhan I just added some readme updates as per your ask. If there's something else you'd like to see, or that I inadvertently missed, please let me know!

File: plugins/outputs/redistimeseries/README.md [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+The RedisTimeSeries output plugin writes metrics to the RedisTimeSeries server.
+
+```toml
+[[outputs.redistimeseries]]
+  ## The address of the RedisTimeSeries server.
+  address = "127.0.0.1:6379"
+
+  ## Password for logging-in to Redis
+  # password = ""
+
```
Comment (srebhan) [inline plugins/outputs/redistimeseries/README.md:12]: Sorry, but this is still not in sync with your `sampleConfig`. I see the following when looking at the code in `plugins/outputs/redistimeseries/redistimeseries.go`

```suggestion
[[outputs.redistimeseries]]
  ## The address of the RedisTimeSeries server.
  address = "127.0.0.1:6379"

  ## Redis ACL credentials
  # username = ""
  # password = ""
  # database = 0
  
  ## Optional TLS Config
  # tls_ca = "/etc/telegraf/ca.pem"
  # tls_cert = "/etc/telegraf/cert.pem"
  # tls_key = "/etc/telegraf/key.pem"
  # insecure_skip_verify = false
```

File: plugins/outputs/redistimeseries/redistimeseries.go (hunk 5 above)
Comment (srebhan) [reply plugins/outputs/redistimeseries/redistimeseries.go:19]: I added a diff, so maybe it becomes more clear on what I'm after. :-)

Comment (chayim) [reply plugins/outputs/redistimeseries/redistimeseries.go:19]: Thank you. I'm new to the project - this is helpful!

File: plugins/outputs/redistimeseries/README.md (hunk 6 above)
Comment (chayim) [reply plugins/outputs/redistimeseries/README.md]: @srebhan I updated the README accordingly (small changes from yours), and similarly added the missing elements to the etc/telegraf.conf file in the correct location. Those items are identical copy+pasta, with each field matching  the _sampleConfig_ const in redistimeseries.go.  

Hopefully this is better ;)

Unfortunately I cannot edit the name of the PR to ensure it matches the Semantic Commit Message format. Can you?

**Review by srebhan — COMMENTED** [review]
@gkorland we are almost there. I would comment the optional options (as this is what we do everywhere else). Furthermore, there is a closing quote missing in your TOML section...

**Review by srebhan — APPROVED** [review]
Looks good to me. Thanks for keep rocking on this PR @gkorland!

**Comment by srebhan** [comment]
@gkorland it seems like there are some merge conflicts. Can you please rebase this PR on the latest master and resolve the conflicts!?

File: etc/telegraf.conf [hunk 7]
```diff
@@ -1631,6 +1631,23 @@
 #   ## Export metric collection time.
 #   # export_timestamp = false
 
+# # Configuration for the RedisTimeSeries server to send metrics to
```
Comment (MyaLongmire) [inline etc/telegraf.conf:1634]: For future reference the telegraf.conf is automatically updated but thank you for covering your bases!

File: plugins/inputs/redis/redis.go [hunk 8]
```diff
@@ -14,8 +14,7 @@ import (
 	"sync"
 	"time"
 
-	"github.com/go-redis/redis"
-
+	"github.com/go-redis/redis/v7"
```
Comment (MyaLongmire) [inline plugins/inputs/redis/redis.go:17]: Is this going to break backward compatibility for before v7?

**Review by MyaLongmire — COMMENTED** [review]
Just one small question and we will be good to go! Thank you for your work on this :)

Can you also run `go mod tidy` and check back in as that should make the tests pass?

**Comment by chayim** [comment]
Pushed!

**Comment by srebhan** [comment]
@chayim and @gkorland I see the tests failing. This is probably due to an update of the redis libraries. Can you maybe take a look?

**Comment by thallredis** [comment]
@chayim can you look at the failing test?

File: plugins/outputs/redistimeseries/redistimeseries.go [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+	Password string          `toml:"password"`
+	Database int             `toml:"database"`
+	Log      telegraf.Logger `toml:"-"`
+	tls.ClientConfig
+	client *redis.Client
+}
+
+func (r *RedisTimeSeries) Connect() error {
+	if r.Address == "" {
+		return fmt.Errorf("Redis address must be specified")
```
Comment (srebhan) [inline plugins/outputs/redistimeseries/redistimeseries.go:40]: Errors must start with lowercase according to Golang convention...
```suggestion
		return errors.New("redis address must be specified")
```

**Review by srebhan — COMMENTED** [review]
Can you please fix the error text? Other than this, the PR looks good.

File: plugins/outputs/redistimeseries/redistimeseries_test.go [hunk 10]
```diff
@@ -0,0 +1,26 @@
+package redistimeseries
```
Comment (powersj) [inline plugins/outputs/redistimeseries/redistimeseries_test.go:1]: As there is zero testing in this PR that would get routinely run, I'd like to see an integration test.

There is an example in the [redis input](https://github.com/influxdata/telegraf/blob/master/plugins/inputs/redis/redis_test.go#L36) using testcontainers that you can copy and paste from.

**Comment by srebhan** [comment]
@gkorland, @chayim can you please fix the CI issue and address the comment of @powersj?

File: plugins/outputs/redistimeseries/redistimeseries_test.go (hunk 10 above)
Comment (chayim) [reply plugins/outputs/redistimeseries/redistimeseries_test.go:1]: @powersj Maybe I misunderstand the testing (and likely)! But doesn't the TestConnectAndWrite in this file, with the MockMetrics cover this use case? I'd like to make sure I'm not missing something - and I'm pretty sure I am.

File: plugins/outputs/redistimeseries/redistimeseries_test.go [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+	}
+
+	// Verify that we can connect to the RedisTimeSeries server
+	err := redis.Connect()
+	require.NoError(t, err)
+
+	// Verify that we can successfully write data to the RedisTimeSeries server
+	err = redis.Write(testutil.MockMetrics())
+	require.NoError(t, err)
+}
```
Comment (srebhan) [inline plugins/outputs/redistimeseries/redistimeseries_test.go:26]: To allow integration tests you want something like this
```suggestion
func TestConnectAndWriteIntegration(t *testing.T) {
	if testing.Short() {
		t.Skip("Skipping integration test in short mode")
	}

	servicePort := "6379"
	container := testutil.Container{
		Image:        "redis:alpine",
		ExposedPorts: []string{servicePort},
		WaitingFor:   wait.ForListeningPort(nat.Port(servicePort)),
	}
	require.NoError(t, container.Start(), "failed to start container")
	defer func() {
		require.NoError(t, container.Terminate(), "terminating container failed")
	}()

	redis := &RedisTimeSeries{
		Address: fmt.Sprintf("%s:%s", container.Address, container.Ports[servicePort]),
	}

	// Verify that we can connect to the RedisTimeSeries server
	require.NoError(t, redis.Connect())

	// Verify that we can successfully write data to the RedisTimeSeries server
	require.NoError(t, redis.Write(testutil.MockMetrics()))
}
```

**Comment by powersj** [comment]
Hi @chayim - @srebhan and I talked and he will write the integration test once this is merged. It is something he wanted to learn anyway.

Can you resolve the test issues and we can land this? Looks like a complaint:

```
vet: plugins/outputs/redistimeseries/redistimeseries.go:40:10: undeclared name: errors
```

You should be able to reproduce locally with `make check`

Thanks!

**Review by powersj — APPROVED** [review]

**Comment by srebhan** [comment]
FYI: PR for the integration test can be found in #11529. 

@gkorland you might also want to change the link in your documentation to the official Telegraf, not that this PR is in... ;-)

**Pull request merged**